src/WebJobs.Extensions.DurableTask/ContextImplementations/DurableEntityContext.cs

```diff
@@ -169,58 +169,86 @@ TState IDurableEntityContext.GetState<TState>(Func<TState> initializer)
         {
             this.ThrowIfInvalidAccess();
 
-            if (!this.StateWasAccessed)
+            if (this.StateWasAccessed)
             {
-                TState defaultValue = default(TState);
+                return (TState)this.CurrentState;
+            }
+
+            TState result;
 
+            if (this.State.EntityExists)
+            {
+                try
+                {
+                    result = JsonConvert.DeserializeObject<TState>(this.State.EntityState);
+                }
+                catch (Exception e)
+                {
+                    throw new EntitySchedulerException($"Failed to deserialize entity state: {e.Message}", e);
+                }
+            }
+            else
+            {
                 if (initializer != null)
                 {
                     try
                     {
-                        defaultValue = initializer();
+                        result = initializer();
                     }
                     catch (Exception e)
                     {
                         throw new EntitySchedulerException($"Failed to initialize entity state: {e.Message}", e);
                     }
                 }
-
-                if (this.State.EntityState != null)
-                {
-                    this.CurrentState = this.GetPopulatedState<TState>(defaultValue, initializer != null);
-                }
                 else
                 {
-                    this.CurrentState = defaultValue;
+                    result = default(TState);
                 }
-
-                this.StateWasAccessed = true;
-                this.State.EntityExists = true;
             }
 
-            return (TState)this.CurrentState;
+            this.CurrentState = result;
+            this.StateWasAccessed = true;
+            this.State.EntityExists = true;
+            return result;
         }
 
-        private TState GetPopulatedState<TState>(TState initialValue, bool usedTypeInitializer)
+        internal TState GetStateWithInjectedDependencies<TState>(Func<TState> constructor)
+            where TState : class
         {
+            this.ThrowIfInvalidAccess();
+
+            if (this.StateWasAccessed)
+            {
+                return (TState)this.CurrentState;
+            }
+
+            TState result;
+
             try
             {
-                if (usedTypeInitializer)
+                result = constructor();
+            }
+            catch (Exception e)
+            {
+                throw new EntitySchedulerException($"Failed to construct entity state object: {e.Message}", e);
+            }
+
+            if (this.State.EntityExists)
```

Comment:
With this approach, we are essentially marking the entity as existing if the customer invokes the entity via `DispatchAsync()` if they are using dependency injection.

Just double checking that this isn't breaking our entity state/existance semantics here.

Reply:
This seems right to me. DispatchAsync() implicitly calls GetState() which always marks the entity as existing. Basically, once the object has been constructed, the entity exists.